scripts/launch-s0-instance.sh

```diff
@@ -3,81 +3,233 @@
 me=$(basename $0)
 dir=$(readlink -f $(dirname $0))
 elastio_aws_id=537513441174
-security_group="elastio-s0-server"
+default_security_group="elastio-s0-server"
 default_instance_type="m5d.2xlarge"
 default_instance_name="$(whoami)-s0-server"
+default_block_size=fixed
+bootstrap=/tmp/s0-bootstrap.yml
+db_mount=/mnt/elastio
+unit_file=/etc/systemd/user/s0.service
+service=s0
+
+create_s0_bootstrap()
+{
+    rm -f $bootstrap
+    cat << EOF > $bootstrap
+#cloud-config
+# Bootstrap a 's0' server on Amazon Linux 2.
+# This script is intended to run as part of the cloud init
+
+write_files:
+  - path: /home/ec2-user/.aws/config
+    owner: ec2-user:ec2-user
+    content: |
+      [default]
+      output = json
+
+  - path: $UNIT_FILE
+    owner: ec2-user:ec2-user
+    permissions: '0664'
+    content: |
+      [Unit]
+      Description="Elastio Scale-to-Zero service"
+      After=network.target
+
+      [Service]
+      User=ec2-user
+      StartLimitInterval=5
+      StartLimitBurst=10
+      TimeoutStopSec=5min
+      ExecStart=/bin/bash -c '/usr/bin/s0 --log-output compact vault serve s3 --bucket $S0_BUCKET --key-id $S0_KMS $BLOCK_SIZE --server-address 0.0.0.0 --db-path $DB_MOUNT 2>&1 | sed -r "s/\x1B\[([0-9]{1,3}(;[0-9]{1,2})?)?[mGK]//g" >> /var/log/s0-service.log'
+      Restart=always
+      RestartSec=5
+
+      [Install]
+      WantedBy=multi-user.target
+
+runcmd:
+  # Format and mount nvme disk
+  - |
+    mkdir -p $DB_MOUNT
+    if [[ -b /dev/nvme1n1 && -b /dev/nvme2n1 ]]; then
+      # Two disks; make an array
+      mdadm --create --verbose /dev/md0 --level=0 --raid-devices=2 /dev/nvme1n1 /dev/nvme2n1
+      mkfs.xfs -f /dev/md0
+      mount /dev/md0 $DB_MOUNT
+    elif [ -b /dev/nvme1n1 ]; then
+      # One disk; no need to involve mdadm
+      parted -a optimal --script /dev/nvme1n1 mklabel gpt
+      parted -a optimal --script /dev/nvme1n1 mkpart primary 0% 100%
+      mkfs.xfs -f /dev/nvme1n1p1
+      mount /dev/nvme1n1p1 $DB_MOUNT
+    fi
+    chown -R ec2-user:ec2-user $DB_MOUNT
+
+  # Configure AWS region
+  - |
+    echo "region = \$(curl -s http://169.254.169.254/latest/dynamic/instance-identity/document | grep -w region | cut -d: -f2 | tr -d '", ')" >> /home/ec2-user/.aws/config
+    cp -r /home/ec2-user/.aws /root/
+
+  # Create s0 vault
```

Comment:
Vault creation and vault serving are separate functions.  The way this is written, every time you launch an instance you will re-create the vault, overwriting any data from a previous run.  That's obviously not particularly useful.

Automating vault creation is now less important than automating vault serving, because Aleks is almost done with elastio/elastio#1181.

Reply:
Will https://github.com/elastio/elastio/pull/1281 be landed before the end of sprint?
What is the new expected behaviour? Will `s0 vault serve s3  ...` create a vault if it doesn't exist? Or is it possible to check, does it exist by some command and create it or not according to the check result?

For now I just know, that if a bucket is empty and a vault doesn't exist, then the service with the command `s0 vault serve s3  ...` will fail with an error and probably just exit returning zero )

Reply:
I don't have a high confidence that Lambda-powered vault creation will land this sprint, no.  But this behavior makes launching the instance useful only as a way to see if ingests are working.  As soon as the instance is launched again, the vault is re-created and all backups are lost.  That's not what we're going for.  I'd rather the user have to manually run `s0 vault create s3 ...` than have to explain this behavior.

BTW the fix for #1375 is awaiting your review )

Reply:
Yup, I agree that the overwriting of the vault on each start is a bad idea. And this behaviour should be changed.
But your suggestion to run manually `s0 vault create s3 ...` buy user is bad idea because it means that they'll need to start s0 service manually then...
I can do the next:
Check existence of the file `vault.json` in the bucket. If it exists, the vault is already created and I don't need to create it here. Otherwise `s0 vault create s3 ...` is running here.
Does it makes sense?

Reply:
Agreed!